sdk/core/core/inc/_az_token_credential.h

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#ifndef AZ_TOKEN_CREDENTIAL_H
+#define AZ_TOKEN_CREDENTIAL_H
+
+#include <az_credential.h>
+#include <az_mut_span.h>
+#include <az_span.h>
+#include <az_span_builder.h>
+
+#include <time.h>
+
+#include <_az_cfg_prefix.h>
+
+enum {
+  _az_TOKEN_CREDENTIAL_TOKEN_BUFFER_SIZE = 3 * (1024 / 2),
+};
+
+typedef struct {
+  az_credential credential;
+  az_span _resource;
+  uint8_t _token_buf[_az_TOKEN_CREDENTIAL_TOKEN_BUFFER_SIZE];
+  size_t _token_size;
+  clock_t _expiration;
```

Comment:
Scopes (here _resource) and access tokens should be stored either in the client or a pipeline policy created by the client rather than on the credential for a number of reasons.

 - Caching these in the credential limits the use of the credential to a single type of client or introduces thread safety issues as @vhvb1989 mentioned in his comments.
 - Caching the token in the client allows the client to react to 401 responses in the case the token has been invalidated before it's expiration.
 - Defining the scopes (or resource) on the credential limits the ability to support services which implement challenged based authentication where the required scopes are passed back in unauthenticated request responses.

Reply:
Agree.
1. I extracted token from the credential
2. Extracting resource will follow
3. Transforming resource to scopes (-> AADv2) will follow

I am only not sure that the client is going to need to handle 401s, I thought that pipeline policies are going to take care of that (for  instance, retry policy is going to work with the auth policy, and tell it to disregard the cached token and request the new one).